ducanhpham0312/zeevision
# feat: instance fetch api

### Linked issue

Closes ducanhpham0312/zeevision-private#119

### Description

- Added `instances` table
- Moved BPMN files to separate `bpmn_resources` table
- Added more fields to `processes` table
- Implemented GraphQL API to allow querying instances through Fetch API.

BPMN files were separated to another table so they can be queried independently and only when they're needed, because of their large (~500 kB) size. Frontend wasn't adjusted to accommodate this leaner strategy (query bpmn file once instead of polling) due to time constraints.

All new internal APIs (excludes GraphQL API) have tests.

NOTE: all records are ordered from newest to oldest in the backend, but table component in the frontend seems to have a bug which causes to reverse them.

### Testing

First ensure that database is not yet initialized from previous runs:

```bash
$ docker compose down --volumes
```

1. Start the test bench
2. Start ZeeVision
3. Open zeevision_db in pgAdmin

Insert the query below to pgAdmin to create a process with associated bpmn resource, and add a instance of that process. Replace `bpmn_file` with base64 encoded BPMN file. You can create the encoded file from one of the files in `zeebe-test-bench/zeebe-workflow-manager/src/main/resources/bpmn` with command:

```bash
$ base64 -w 0 <file>
```

The queries to run:

```sql
INSERT INTO processes(process_definition_key, bpmn_process_id, version, deployment_time)
VALUES(123, 'hello-world', 1, NOW());

INSERT INTO bpmn_resources(bpmn_process_id, bpmn_file)
VALUES('hello-world', '<the bpmn file>');

INSERT INTO instances(process_instance_key, process_definition_key, version, status, start_time)
VALUES(1, 123, 1, 'ACTIVE', NOW());
```

Check that processes page has one process. It also should have visible BPMN figure, and instance in its instances list.
Check that instances page has the same instance that the process had.
